Trey/signals docs (#3298)
TobikoData/sqlmesh: docs/guides/signals.md

  kind FULL,
  signals [
    (threshold = 0.5), # specify threshold value
  ]
);

SELECT 1
```

The next time this project is `sqlmesh run`, our signal will metaphorically flip a coin to determine whether the model should be evaluated.

### Advanced Example

Multiple signals can be used within one model. A signal can return a subset of the intervals within a batch. The intersection of the intervals returned by all the signals determines what is evaluated.

In this example, there are two signals.

```python
import typing as t
from datetime import datetime

from sqlmesh.core.scheduler import signal_factory, Batch, Signal
from sqlmesh.utils.date import to_datetime

class AlwaysReady(Signal):
    # signal that indicates every interval is always ready
    def check_intervals(self, batch: Batch) -> t.Union[bool, Batch]:
        return True

class OneweekAgo(Signal):
    def __init__(self, dt: datetime):
        self.dt = dt

    # signal that returns only intervals that are <= 1 week ago
    def check_intervals(self, batch: Batch) -> t.Union[bool, Batch]:
        return [
            (start, end)
            for start, end in batch
            if start <= self.dt
        ]
```

> 

Reviewer: ```suggestion
Instead of returning a single `True`/`False` value for whether a batch of intervals is ready for evaluation, the `OneweekAgo` signal class returns specific intervals from the batch.

Its `check_intervals` method accepts a `dt` datetime argument, to which It compares the beginning of each interval in the batch. If the interval start is before that argument, the interval is ready for evaluation and included in the returned list.
```

> These signals can be added to a model like so. `Kind` is used so that the signal factory can instantiate the correct one. In this case, kind is an arbitrary key value which serves as an example of how to handle multiple signals.

Reviewer: ```suggestion
These signals can be added to a model like so. Now that we have more than one signal, we must have a way to tell the signal factory which signal should be called.

In this example, we use the `kind` key to tell the signal factory which signal class should be called. The key name is arbitrary, and you may choose any key name you want.

This example model specifies two `kind` values so both signals are called.
```

```sql linenums="1" hl_lines="7-10"
MODEL (
  name example.signal_model,
  kind INCREMENTAL_BY_TIME_RANGE (
    time_column ds,
  ),
  start '2 week ago',
  signals [
    (kind = 'a'),
    (kind = 'b'),
  ]
);

SELECT @start_ds AS ds
```

> The appropriate signal can then be instantiated through the "kind" parameter.

Reviewer: ```suggestion
Our signal factory definition extracts the `kind` key value from the `signal_metadata`, then instantiates the signal class corresponding to the value:
```
